File: src/main/java/com/vjiki/music/mapper/PlaylistSongMapper.java
package com.vjiki.music.mapper;

import com.vjiki.music.dto.PlaylistSongResponse;
import com.vjiki.music.entity.PlaylistSong;

public final class PlaylistSongMapper {

    private PlaylistSongMapper() {
    }

    public static PlaylistSongResponse toResponse(PlaylistSong playlistSong) {
        return PlaylistSongResponse.builder()
                .id(playlistSong.getId())
                .playlistId(playlistSong.getPlaylist() == null ? null : playlistSong.getPlaylist().getId())
                .songId(playlistSong.getSong() == null ? null : playlistSong.getSong().getId())
                .songTitle(playlistSong.getSong() == null ? null : playlistSong.getSong().getTitle())
                .songArtist(playlistSong.getSong() == null
                        ? null
                        : SongMapper.extractArtist(playlistSong.getSong().getArtists()))
                .songAudioUrl(playlistSong.getSong() == null
                        ? null
                        : SongMapper.extractUrl(playlistSong.getSong().getAudioUrls()))
                .songCoverUrl(playlistSong.getSong() == null
                        ? null
                        : SongMapper.extractUrl(playlistSong.getSong().getCoverUrls()))
                .position(playlistSong.getPosition())
                .addedAt(playlistSong.getAddedAt())
                .addedBy(playlistSong.getAddedBy() == null ? null : playlistSong.getAddedBy().getId())
                .build();
    }
}

File: src/main/kotlin/com/vjiki/music/mapper/PlaylistSongMapper.kt
package com.vjiki.music.mapper

import com.vjiki.music.dto.PlaylistSongResponse
import com.vjiki.music.entity.PlaylistSong
import org.springframework.stereotype.Component

@Component
object PlaylistSongMapper {
    fun PlaylistSong.toResponse(): PlaylistSongResponse {
        return PlaylistSongResponse(
            id = id,
            playlistId = playlist.id,
            songId = song.id,
            songTitle = song.title,
            songArtist = extractArtist(song.artists),
            songAudioUrl = extractUrl(song.audioUrls),
            songCoverUrl = extractUrl(song.coverUrls),
            position = position,
            addedAt = addedAt,
            addedBy = addedBy?.id
        )
    }

    private fun extractArtist(artists: Map<String, List<String>>?): String? {
        return artists?.get("default")?.firstOrNull()
    }

    private fun extractUrl(urls: Map<String, String>?): String? {
        return urls?.get("default")
    }
}

